File: app/src/main/java/com/kota/Bahamut/Dialogs/Dialog_PaintColor.java
package com.kota.Bahamut.Dialogs;

import static com.kota.Bahamut.Service.CommonFunctions.getContextString;

import android.text.SpannableString;
import android.text.Spanned;
import android.text.style.BackgroundColorSpan;
import android.text.style.ForegroundColorSpan;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.LinearLayout;
import android.widget.Spinner;
import android.widget.TextView;

import com.kota.ASFramework.Dialog.ASDialog;
import com.kota.Bahamut.R;
import com.kota.Telnet.Reference.TelnetAnsiCode;

import java.util.Objects;

public class Dialog_PaintColor extends ASDialog implements View.OnClickListener {
    LinearLayout mainLayout;
    boolean _isRecovery = true;
    boolean _isHighlight = false;
    int _frontColor = 0;
    int _backColor = 0;
    String _outputParam;
    Dialog_PaintColor_Listener _listener;
    CheckBox _recovery_box;
    CheckBox _highlight_box;
    Spinner _front_color_spinner;
    Spinner _back_color_spinner;
    TextView _textview_param;
    TextView _textview_sample;

    Button _send_button;
    Button _cancel_button;

    CompoundButton.OnCheckedChangeListener _recovery_listener = (buttonView, isChecked) -> {
        _isRecovery = isChecked;
        if (isChecked) {
            _frontColor = -1;
            _backColor = -1;
            _isHighlight = false;
            _front_color_spinner.setSelection(_frontColor);
            _back_color_spinner.setSelection(_backColor);
            _highlight_box.setChecked(_isHighlight);

        }
        generateOutputParam();
    };
    CompoundButton.OnCheckedChangeListener _highlight_listener = (buttonView, isChecked) -> {
        _isHighlight = isChecked;
        if (_isRecovery) {
            _isRecovery = false;
            _recovery_box.setChecked(false);
        }
        generateOutputParam();
    };

    AdapterView.OnItemSelectedListener _frontColor_listener = new AdapterView.OnItemSelectedListener() {
        @Override
        public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
            _frontColor = i;
            if (i>0 && _isRecovery) {
                _isRecovery = false;
                _recovery_box.setChecked(false);
            }
            generateOutputParam();
        }

        @Override
        public void onNothingSelected(AdapterView<?> adapterView) {
            _frontColor = -1;
        }
    };

    AdapterView.OnItemSelectedListener _backColor_listener = new AdapterView.OnItemSelectedListener() {
        @Override
        public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
            _backColor = i;
            if (i>0 && _isRecovery) {
                _isRecovery = false;
                _recovery_box.setChecked(false);
            }
            generateOutputParam();
        }

        @Override
        public void onNothingSelected(AdapterView<?> adapterView) {
            _backColor = -1;
        }
    };

    public String getName() {
        return "BahamutPostArticlePaintColor";
    }

    public Dialog_PaintColor() {
        requestWindowFeature(1);
        setContentView(R.layout.dialog_paint_color);
        if (getWindow()!=null)
            getWindow().setBackgroundDrawable(null);
        setTitle(getContextString(R.string.post_article_page_paint_color));
        mainLayout = findViewById(R.id.dialog_paint_color_content_view);

        // 還原
        _recovery_box = mainLayout.findViewById(R.id.dialog_paint_color_check_recovery);
        _recovery_box.setOnCheckedChangeListener(_recovery_listener);
        mainLayout.findViewById(R.id.dialog_paint_color_check_recovery_item).setOnClickListener(view -> _recovery_box.setChecked(!_recovery_box.isChecked()));

        // 前景
        ArrayAdapter<String> adapter_front_color = new ArrayAdapter<>(getContext(), R.layout.simple_spinner_item, getContext().getResources().getStringArray(R.array.dialog_paint_color_items));
        adapter_front_color.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        _front_color_spinner = mainLayout.findViewById(R.id.post_article_page_paint_color_front_spinner);
        _front_color_spinner.setAdapter(adapter_front_color);
        _front_color_spinner.setOnItemSelectedListener(_frontColor_listener);

        // 背景
        ArrayAdapter<String> adapter_back_color = new ArrayAdapter<>(getContext(), R.layout.simple_spinner_item, getContext().getResources().getStringArray(R.array.dialog_paint_color_items));
        adapter_back_color.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        _back_color_spinner = mainLayout.findViewById(R.id.post_article_page_paint_color_back_spinner);
        _back_color_spinner.setAdapter(adapter_back_color);
        _back_color_spinner.setOnItemSelectedListener(_backColor_listener);

        // 亮色
        _highlight_box = mainLayout.findViewById(R.id.dialog_paint_color_check_highlight);
        _highlight_box.setOnCheckedChangeListener(_highlight_listener);
        mainLayout.findViewById(R.id.dialog_paint_color_check_highlight_item).setOnClickListener(view -> _highlight_box.setChecked(!_highlight_box.isChecked()));

        // 文字
        _textview_param = mainLayout.findViewById(R.id.dialog_paint_color_param);
        _textview_sample = mainLayout.findViewById(R.id.dialog_paint_color_sample);

        // 按鈕
        _send_button = mainLayout.findViewById(R.id.send);
        _send_button.setOnClickListener(this);
        _cancel_button = mainLayout.findViewById(R.id.cancel);
        _cancel_button.setOnClickListener(this);

        setDialogWidth();
    }

    void generateOutputParam() {
        if (_isRecovery) {
            _outputParam = "*[m";
        } else {
            _outputParam = "*[";
            if (_isHighlight)
                _outputParam+="1;";
            if (_frontColor>0) {
                _outputParam += "3" + (_frontColor-1);
                if (_backColor>0) {
                    _outputParam += ";4" + (_backColor-1);
                }
            } else
            if (_backColor>0) {
                _outputParam += "4" + (_backColor-1);
            }
            _outputParam += "m";
        }
        // 內容改變
        _textview_param.setText(_outputParam);

        // 顏色改變
        SpannableString _sample = new SpannableString(_textview_sample.getText());
        Object[] spansToRemove = _sample.getSpans(0, _sample.length(), Object.class);

        if (_isRecovery) {
            // 還原
            for (Object span:spansToRemove) {
                if (span instanceof ForegroundColorSpan || span instanceof BackgroundColorSpan)
                    _sample.removeSpan(span);
            }
        } else {
            byte paintColor;
            if (_frontColor>0) {
                paintColor = (byte) (_frontColor-1);
                if (_isHighlight)
                    paintColor+=8;
                ForegroundColorSpan colorSpan = new ForegroundColorSpan(TelnetAnsiCode.getTextColor(paintColor));
                _sample.setSpan(colorSpan, 0, _sample.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
            } else {
                for (Object span:spansToRemove) {
                    if (span instanceof ForegroundColorSpan)
                        _sample.removeSpan(span);
                }
            }

            if (_backColor>0) {
                paintColor = (byte) (_backColor-1);
                BackgroundColorSpan colorSpan = new BackgroundColorSpan(TelnetAnsiCode.getBackgroundColor(paintColor));
                _sample.setSpan(colorSpan, 0, _sample.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
            } else {
                for (Object span:spansToRemove) {
                    if (span instanceof BackgroundColorSpan)
                        _sample.removeSpan(span);
                }
            }
        }
        _textview_sample.setText(_sample);
    }

    public void setListener(Dialog_PaintColor_Listener listener) {
        _listener = listener;
    }

    @Override
    public void onClick(View view) {
        if (view == _send_button && _listener != null) {
            _listener.onPaintColorDone(_outputParam);
        }
        dismiss();
    }
    // 變更dialog寬度
    void setDialogWidth() {
        int screenWidth = getContext().getResources().getDisplayMetrics().widthPixels;
        int dialog_width = (int) (screenWidth*0.7);
        ViewGroup.LayoutParams oldLayoutParams = mainLayout.getLayoutParams();
        oldLayoutParams.width = dialog_width;
        mainLayout.setLayoutParams(oldLayoutParams);
    }
}

File: app/src/main/java/com/kota/Bahamut/dialogs/DialogPaintColor.kt
package com.kota.Bahamut.dialogs

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import com.kota.Bahamut.R
import com.kota.Bahamut.service.CommonFunctions
import com.kota.Bahamut.ui.components.BahaCheckbox
import com.kota.Bahamut.ui.components.BahaDropdownMenu
import com.kota.Bahamut.ui.components.BahaText
import com.kota.Bahamut.ui.components.ButtonType
import com.kota.Bahamut.ui.dialogs.BahaAlertDialogContent
import com.kota.Bahamut.ui.dialogs.BahaDialogButton
import com.kota.Bahamut.ui.theme.AppTheme
import com.kota.asFramework.dialog.ASDialog
import com.kota.telnet.reference.TelnetAnsiCode

class DialogPaintColor : ASDialog() {
    private var dialogPaintColorListener: DialogPaintColorListener? = null

    override val name: String?
        get() = "BahamutPostArticlePaintColor"

    init {
        setTitle(CommonFunctions.getContextString(R.string.post_article_page_paint_color))
        setComposeContent {
            Content()
        }
    }

    @Composable
    private fun Content() {
        val context = LocalContext.current
        val colors = AppTheme.colors
        val scrollState = rememberScrollState()

        val colorOptions = remember {
            context.resources.getStringArray(R.array.dialog_paint_color_items)
        }

        var isRecovery by remember { mutableStateOf(true) }
        var isHighlight by remember { mutableStateOf(false) }
        var frontColor by remember { mutableIntStateOf(0) }
        var backColor by remember { mutableIntStateOf(0) }

        var frontExpanded by remember { mutableStateOf(false) }
        var backExpanded by remember { mutableStateOf(false) }

        // Generate outputParam
        val outputParam = remember(isRecovery, isHighlight, frontColor, backColor) {
            if (isRecovery) {
                "*[m"
            } else {
                val sb = StringBuilder("*[")
                if (isHighlight) sb.append("1;")
                if (frontColor > 0) {
                    sb.append("3").append(frontColor - 1)
                    if (backColor > 0) {
                        sb.append(";4").append(backColor - 1)
                    }
                } else if (backColor > 0) {
                    sb.append("4").append(backColor - 1)
                }
                sb.append("m")
                sb.toString()
            }
        }

        // Compute preview textColor and bgColor
        val sampleTextColor = remember(isRecovery, isHighlight, frontColor) {
            if (isRecovery || frontColor <= 0) {
                colors.textPrimary
            } else {
                var paintColor: Byte = (frontColor - 1).toByte()
                if (isHighlight) paintColor = (paintColor + 8).toByte()
                Color(TelnetAnsiCode.getTextColor(paintColor))
            }
        }

        val sampleBgColor = remember(isRecovery, backColor) {
            if (isRecovery || backColor <= 0) {
                Color.Transparent
            } else {
                val paintColor: Byte = (backColor - 1).toByte()
                Color(TelnetAnsiCode.getBackgroundColor(paintColor))
            }
        }

        BahaAlertDialogContent(
            title = stringResource(R.string.post_article_page_paint_color),
            buttons = listOf(
                BahaDialogButton(
                    text = stringResource(R.string.cancel),
                    type = ButtonType.SECONDARY,
                    onClick = { dismiss() }
                ),
                BahaDialogButton(
                    text = stringResource(R.string.send),
                    type = ButtonType.NORMAL,
                    onClick = {
                        dialogPaintColorListener?.onPaintColorDone(outputParam)
                        dismiss()
                    }
                )
            )
        ) {
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .verticalScroll(scrollState)
            ) {
                // 還原 Checkbox (文字在左，核取方塊在右)
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .clickable {
                            val newRecovery = !isRecovery
                            isRecovery = newRecovery
                            if (newRecovery) {
                                frontColor = 0
                                backColor = 0
                                isHighlight = false
                            }
                        }
                        .padding(vertical = 6.dp),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    BahaText(
                        text = stringResource(R.string.post_article_page_paint_color_recovery),
                        color = colors.textPrimary
                    )
                    Spacer(modifier = Modifier.width(8.dp))
                    BahaText(
                        text = stringResource(R.string.post_article_page_paint_color_recovery2),
                        color = colors.textSecondary
                    )
                    Spacer(modifier = Modifier.weight(1f))
                    BahaCheckbox(
                        checked = isRecovery,
                        onCheckedChange = { isChecked ->
                            isRecovery = isChecked
                            if (isChecked) {
                                frontColor = 0
                                backColor = 0
                                isHighlight = false
                            }
                        }
                    )
                }

                Spacer(modifier = Modifier.height(4.dp))

                // 前景色
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .clickable { frontExpanded = true }
                        .padding(vertical = 6.dp),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    BahaText(
                        text = stringResource(R.string.post_article_page_paint_color_front),
                        color = colors.textPrimary
                    )
                    Spacer(modifier = Modifier.width(8.dp))
                    BahaText(
                        text = stringResource(R.string.post_article_page_paint_color_front2),
                        color = colors.textSecondary
                    )
                    Spacer(modifier = Modifier.weight(1f))
                    Box {
                        Row(
                            verticalAlignment = Alignment.CenterVertically,
                            modifier = Modifier.padding(horizontal = 8.dp, vertical = 4.dp)
                        ) {
                            BahaText(
                                text = colorOptions.getOrElse(frontColor) { "" },
                                color = colors.textPrimary,
                                fontSize = AppTheme.fontSize.subtitle
                            )
                            Spacer(modifier = Modifier.width(6.dp))
                            BahaText(
                                text = "▾",
                                color = colors.textSecondary,
                                fontSize = AppTheme.fontSize.subtitle
                            )
                        }
                        BahaDropdownMenu(
                            expanded = frontExpanded,
                            onDismissRequest = { frontExpanded = false },
                            items = colorOptions,
                            selectedIndex = frontColor,
                            onItemSelected = { index ->
                                frontColor = index
                                if (index > 0 && isRecovery) isRecovery = false
                            },
                            fontSize = AppTheme.fontSize.subtitle
                        )
                    }
                }

                Spacer(modifier = Modifier.height(4.dp))

                // 背景色
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .clickable { backExpanded = true }
                        .padding(vertical = 6.dp),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    BahaText(
                        text = stringResource(R.string.post_article_page_paint_color_back),
                        color = colors.textPrimary
                    )
                    Spacer(modifier = Modifier.width(8.dp))
                    BahaText(
                        text = stringResource(R.string.post_article_page_paint_color_back2),
                        color = colors.textSecondary
                    )
                    Spacer(modifier = Modifier.weight(1f))
                    Box {
                        Row(
                            verticalAlignment = Alignment.CenterVertically,
                            modifier = Modifier.padding(horizontal = 8.dp, vertical = 4.dp)
                        ) {
                            BahaText(
                                text = colorOptions.getOrElse(backColor) { "" },
                                color = colors.textPrimary,
                                fontSize = AppTheme.fontSize.subtitle
                            )
                            Spacer(modifier = Modifier.width(6.dp))
                            BahaText(
                                text = "▾",
                                color = colors.textSecondary,
                                fontSize = AppTheme.fontSize.subtitle
                            )
                        }
                        BahaDropdownMenu(
                            expanded = backExpanded,
                            onDismissRequest = { backExpanded = false },
                            items = colorOptions,
                            selectedIndex = backColor,
                            onItemSelected = { index ->
                                backColor = index
                                if (index > 0 && isRecovery) isRecovery = false
                            },
                            fontSize = AppTheme.fontSize.subtitle
                        )
                    }
                }

                Spacer(modifier = Modifier.height(4.dp))

                // 亮色 Checkbox
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .clickable {
                            val newHighlight = !isHighlight
                            isHighlight = newHighlight
                            if (newHighlight && isRecovery) isRecovery = false
                        }
                        .padding(vertical = 6.dp),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    BahaText(
                        text = stringResource(R.string.post_article_page_paint_color_highlight),
                        color = colors.textPrimary
                    )
                    Spacer(modifier = Modifier.width(8.dp))
                    BahaText(
                        text = stringResource(R.string.post_article_page_paint_color_highlight2),
                        color = colors.textSecondary
                    )
                    Spacer(modifier = Modifier.weight(1f))
                    BahaCheckbox(
                        checked = isHighlight,
                        onCheckedChange = { isChecked ->
                            isHighlight = isChecked
                            if (isChecked && isRecovery) isRecovery = false
                        }
                    )
                }

                Spacer(modifier = Modifier.height(10.dp))

                // 預覽參數與預覽樣本
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(vertical = 8.dp),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    BahaText(
                        text = outputParam,
                        color = colors.textSecondary
                    )
                    Box(
                        modifier = Modifier
                            .background(sampleBgColor)
                            .padding(horizontal = 10.dp, vertical = 4.dp)
                    ) {
                        BahaText(
                            text = stringResource(R.string.dialog_paint_color_sample_ch),
                            color = sampleTextColor
                        )
                    }
                }
            }
        }
    }

    fun setListener(listener: DialogPaintColorListener?) {
        this.dialogPaintColorListener = listener
    }
}
